### DISCUSSION TITLE
Creating an InstrumentationModule that requires external library

### DISCUSSION THREAD

**Discussion by oxeye-mher** [open-telemetry/opentelemetry-java-instrumentation#9126] [Q&A]
Hello everyone,

I'm currently working on an `InstrumentationModule` that needs to be loaded as an `Extension`. One of my `AdviceMethod`s is responsible for creating a JSON text, and I'd like to utilize the `com.google.gson.Gson` Java library (or any other suitable library) for this purpose.

The `build.gradle` file contains the following line: `implementation("com.google.code.gson:gson:2.10.1")`, ensuring that the `com.google.gson` library exists within the `extension.jar`.

However, I've encountered an issue. I'm struggling to make Opentelemetry inject the `Gson` library into the application code. After reading the documentation, I tried using the `getAdditionalHelperClassNames` function. Unfortunately, since `com.google.gson` comprises multiple classes, using `getAdditionalHelperClassNames` would require listing all Gson classes in the correct order, which I believe is not the optimal solution.

Is there a way to inject the entire `com.google.gson` library into the application code so that I can use it without explicit instrumentation? 

I did manage to make it work by adding the `com.google.com` library to the `opentelemetry-javaagent`. However, this solution isn't ideal for me as I prefer to use the original `opentelemetry-javaagent` and add my code only as an extension.

Any help would be greatly appreciated. Thank you!

**Comment by laurit** [comment]
I think you should be using `isHelperClass` instead of `getAdditionalHelperClassNames`. With `isHelperClass` you don't need to list full class names and can just use `startsWith` and don't need to worry about the order.
Injecting large libraries has helper classes isn't usually a good idea. If the application also contains `gson` you'd have to shade your copy of `gson` to avoid clashing with it. For large libraries it is possible that the amount of muzzle (our safety checker that determines whether instrumentation can be safely applied) collects so much info that the method generated based on it exceeds max method size in class file. Instead of an extension you may need to package your code as a custom distro and instead of injecting `gson` as helper classes add it to agent class loader and have an api in boot loader that lets you use `gson` in agent from application code.

**Reply by oxeye-mher** [reply]
Edit - I notice now that when I use the demo extension, `IsHelperClass` does exactly what you mentioned, but when I override `IsHelperClass` in my instrumentation class - is never called, any ideas where to look for an issue? 


Thanks for the answer!
I tried to use `isHelperClass` but it didn't work.
but I managed to make it work using creating own `javaagent` distro.
But what will happen in case there is already `Gson` library in the instrumented code and I inject my `javaagent` that contains my `Gson`  ? Should I do something special or `javaagent` is dealing with making the injection be safe?

**Reply by mateuszrzeszutek** [reply]
> I tried to use isHelperClass but it didn't work.

Are you using the [muzzle gradle plugin](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/examples/extension/build.gradle#L16)? It does some magic during compile time and registers the `isHelperClass` classes as helpers.

> But what will happen in case there is already `Gson` library in the instrumented code and I inject my `javaagent` that contains my `Gson` ? Should I do something special or `javaagent` is dealing with making the injection be safe?

As Lauri mentioned, you should probably shade your `gson` copy. If the app contains an incompatible `gson` version the instrumentation might just not work.

**Reply by oxeye-mher** [reply]
You're right, I missed that I need to use muzzle for that, thank you very much! I will add muzzle plugin

**Reply by Henrique-rr-Martins** [reply]
Hello guys. Right now I am trying to create an extension from Opentelemtry JavaAgente but I have external libraries and I am using Maven with a pom.xml. I don't have experience with gradle. How could I reach the same result using Maven? is it possible?
